robot/packages/cloud_package_manger.go

```diff
@@ -0,0 +1,514 @@
+package packages
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"hash"
+	"hash/crc32"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/edaniels/golog"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+	pb "go.viam.com/api/app/packages/v1"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/config"
+)
+
+const (
+	allowedContentType = "application/x-gzip"
+)
+
+var (
+	_ Manager        = (*cloudManager)(nil)
+	_ ManagerAndSync = (*cloudManager)(nil)
+)
+
+type managedPackage struct {
+	thePackage config.PackageConfig
+	modtime    time.Time
+}
+
+type cloudManager struct {
+	client          pb.PackageServiceClient
+	httpClient      http.Client
+	packagesDataDir string
+	packagesDir     string
+
+	managedPackages map[PackageName]*managedPackage
+	mu              sync.RWMutex
+
+	logger golog.Logger
+}
+
+// NewCloudManager creates a new manager with the given package service client and directory to sync to.
+func NewCloudManager(client pb.PackageServiceClient, viamDir string, logger golog.Logger) (ManagerAndSync, error) {
+	packagesDir := filepath.Join(viamDir, "packages")
+	packagesDataDir := filepath.Join(packagesDir, ".data")
+
+	if err := os.MkdirAll(packagesDir, os.ModePerm); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(packagesDataDir, os.ModePerm); err != nil {
+		return nil, err
+	}
+
+	return &cloudManager{
+		client:          client,
+		httpClient:      http.Client{Timeout: time.Minute * 30},
+		packagesDir:     packagesDir,
+		packagesDataDir: packagesDataDir,
+		managedPackages: make(map[PackageName]*managedPackage),
+		logger:          logger.Named("package_manager"),
+	}, nil
+}
+
+// PackagePath returns the package if it exists and already download. If it does not exist it returns a ErrPackageMissing error.
+//
+// TODO(adam): test.
+func (m *cloudManager) PackagePath(name PackageName) (string, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	p, ok := m.managedPackages[name]
+	if !ok {
+		return "", ErrPackageMissing
+	}
+
+	return m.localNamedPath(p.thePackage), nil
+}
+
+// TODO(adam): test.
+func (m *cloudManager) RefPath(refPath string) (string, error) {
+	ref := config.HasPackageReference(refPath)
+
+	// If no reference just return original path.
+	if ref == nil {
+		return refPath, nil
+	}
+
+	packagePath, err := m.PackagePath(PackageName(ref.Package))
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(packagePath, path.Join("/", ref.PathInPackage)), nil
```

Comment:
I cleaned this up a bit. My intention was to disallow any path traversal with `${packages.some-package}/../../other-dir`. This does fix it because path.Join() calls Clean() at the end.

Cleaned up with just `path.Join(packagePath, path.Clean(ref.PathInPackage))`